cin_validator/rules/cin2022_23/rule_8940.py

```diff
@@ -0,0 +1,262 @@
+    df_merged["ERROR_ID"] = tuple(
+        zip(df_merged[LAchildID], df_merged["CPPID_cpp"], df_merged["CPPID_cpp2"])
+    )
+    print("\n df_merged zipped\n", df_merged)
+    # The merges were done on copies of cpp_df and reviews_df so that the column names in dataframes themselves aren't affected by the suffixes.
+    # we can now map the suffixes columns to their corresponding source tables such that the failing ROW_IDs and ERROR_IDs exist per table.
+    df_cpp_issues = (
+        df_cpp.merge(df_merged, left_on="ROW_ID", right_on="ROW_ID_cpp")
+        .groupby("ERROR_ID", group_keys=False)["ROW_ID"]
+        .apply(list)
+        .reset_index()
+    )
+    df_cpp2_issues = (
+        df_cpp2.merge(df_merged, left_on="ROW_ID", right_on="ROW_ID_cpp2")
+        .groupby("ERROR_ID", group_keys=False)["ROW_ID"]
+        .apply(list)
+        .reset_index()
+    )
+    print("\n df_cpp_issues\n", df_cpp_issues)
+    print("\n df_cpp2_issues\n", df_cpp2_issues)
+
+    # Ensure that you maintain the ROW_ID, and ERROR_ID column names which are shown above. They are keywords in this project.
+    rule_context.push_type_2(
+        table=ChildProtectionPlans, columns=[CPPstartDate], row_df=df_cpp_issues
+    )
+    rule_context.push_type_2(
+        table=ChildProtectionPlans,
+        columns=[CPPstartDate, CPPendDate],
+        row_df=df_cpp2_issues,
+    )
+
+
+def test_validate():
+    # Create some sample data such that some values pass the validation and some fail.
+    sample_header = pd.DataFrame(
+        [{ReferenceDate: "31/03/2001"}]  # the census start date here will be 01/04/2000
+    )
+
+    sample_cpp = pd.DataFrame(
+        [
+            {
+                "LAchildID": "child1",  # 0 Pass
+                "CPPstartDate": "26/05/2000",
+                "CPPendDate": "26/10/2000",
+                "CPPID": "cinID1",
+            },
+            {
+                "LAchildID": "child1",  # 1 Fail
+                "CPPstartDate": "26/08/2000",
+                "CPPendDate": "26/12/2000",
+                "CPPID": "cinID12",
+            },
+            {
+                "LAchildID": "child2",  # 2 Pass
+                "CPPstartDate": "26/05/2000",
+                "CPPendDate": "26/10/2000",
+                "CPPID": "cinID2",
+            },
+            {
+                "LAchildID": "child2",  # 3 Pass
+                "CPPstartDate": "26/10/2000",
+                "CPPendDate": "26/12/2000",
+                "CPPID": "cinID22",
+            },
+            {
+                "LAchildID": "child3",  # 4 Pass
+                "CPPstartDate": "26/05/2000",
+                "CPPendDate": pd.NA,
+                "CPPID": "cinID3",
+            },
+            {
+                "LAchildID": "child3",  # 5 Fail
+                "CPPstartDate": "26/08/2000",
+                "CPPendDate": "26/10/2000",
+                "CPPID": "cinID32",
+            },
+        ]
+    )
+
```

Comment:
Niceeee
Full test coverage !

Reply:
The one thing I'm not sure about is, should type 2 rules test ALL outputs? 

E.g. this returns errors in two different tables but the tests only make sure that one of the tables has output as expected (as in the base rule we are advised to copy).

assert issue_table
assert issue_columns
assert issue_rows

Reply:
Yes. 
Ideally, we would check issue_table, issue_columns, and issue_rows for all the tables created in the push. 

Choosing one to check is good but checking all is better (though longer).